presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TableLayoutRewriter.java

```diff
@@ -97,4 +104,14 @@ public PlanNode planTableScan(TableScanNode node, Expression predicate)
 
         return result;
     }
+
+    private boolean hasAllNeededOutputs(TableLayoutResult layout, TableScanNode node)
+    {
+        List<ColumnHandle> nodeColumnHandles = node.getOutputSymbols().stream()
+                .map(node.getAssignments()::get)
+                .collect(toImmutableList());
+        return layout.getLayout().getColumns()
+                .map(columnHandles -> columnHandles.containsAll(nodeColumnHandles))
+                .orElse(true);
```

Comment:
IMO starnge usage of `Optional` in `layout.getLayout().getColumns()` makes this unclear. Change to:
```
if (!layout.getLayout().getColumns().isPresent()) {
    return true;
}

List<ColumnHandle> nodeColumnHandles = node.getOutputSymbols().stream()
                 .map(node.getAssignments()::get)
                 .collect(toImmutableList());

return layout.getLayout().getColumns().containsAll(nodeColumnHandles);
```

Also, this gonna perform quadratic (think of table having 1,000 columns), so use sets.

(Also, in the unlikely case there are 2+ layouts, the conversion `node` to `nodeColumnHandles` will be repeated. Maybe that's not a big problem, since you'd need a set on the left side.)

Reply:
If `l` is a list and `c` is any collection, then `l.containsAll(c)` is `0(l.size() * c.size())`.